File: src/pages/api/scan.ts
import type { NextApiRequest, NextApiResponse } from 'next'
import http from 'node:http'
import https from 'node:https'
import tls from 'node:tls'

import type { ScanResult, Verdict } from '@/lib/types'

const SECURITY_HEADERS = [
  'Strict-Transport-Security',
  'Content-Security-Policy',
  'X-Frame-Options',
  'X-Content-Type-Options',
  'Referrer-Policy',
  'Permissions-Policy',
]

const MAX_REDIRECTS = 10
const REQUEST_TIMEOUT_MS = 8000

type ApiResponse =
  | { results: ScanResult[] }
  | { error: string }

type ProbeResult = {
  statusCode: number | null
  finalUrl: string | null
  redirectChain: string[]
  responseTimeMs: number | null
  headers: Record<string, string | string[] | undefined>
}

export default async function handler(
  req: NextApiRequest,
  res: NextApiResponse<ApiResponse>,
) {
  if (req.method === 'OPTIONS') {
    res.setHeader('Allow', 'POST, OPTIONS')
    return res.status(204).end()
  }

  if (req.method !== 'POST') {
    res.setHeader('Allow', 'POST, OPTIONS')
    return res.status(405).json({ error: 'POST only' })
  }

  try {
    const urls: unknown[] = Array.isArray(req.body?.urls) ? req.body.urls : []
    if (!urls.length) {
      return res.status(400).json({ error: 'No URLs provided' })
    }

    const results = await Promise.all(
      urls.slice(0, 10).map((url: unknown) => scanUrl(String(url))),
    )

    return res.status(200).json({ results })
  } catch (error) {
    const message = error instanceof Error ? error.message : 'Unknown error'
    return res.status(500).json({ error: message })
  }
}

async function scanUrl(rawUrl: string): Promise<ScanResult> {
  const result: ScanResult = {
    url: rawUrl,
    valid: false,
    status_code: null,
    final_url: null,
    redirect_chain: [],
    response_time_ms: null,
    ssl_valid: null,
    ssl_days_left: null,
    ssl_issuer: null,
    missing_headers: [],
    present_headers: [],
    verdict: 'ERROR',
    error: null,
  }

  const validation = normalizeUrl(rawUrl)
  if (!validation.valid) {
    result.error = validation.error
    return result
  }

  result.valid = true
  result.url = validation.url

  try {
    const probe = await httpProbe(validation.url)
    result.status_code = probe.statusCode
    result.final_url = probe.finalUrl
    result.redirect_chain = probe.redirectChain
    result.response_time_ms = probe.responseTimeMs

    const headerNames = new Set(Object.keys(probe.headers).map((key) => key.toLowerCase()))
    for (const header of SECURITY_HEADERS) {
      if (headerNames.has(header.toLowerCase())) {
        result.present_headers.push(header)
      } else {
        result.missing_headers.push(header)
      }
    }

    if (result.final_url?.startsWith('https://')) {
      const sslInfo = await sslCheck(result.final_url)
      result.ssl_valid = sslInfo.sslValid
      result.ssl_days_left = sslInfo.sslDaysLeft
      result.ssl_issuer = sslInfo.sslIssuer
      if (sslInfo.error) {
        result.error = sslInfo.error
      }
    }

    result.verdict = classifyVerdict(result)
  } catch (error) {
    result.error = error instanceof Error ? error.message : 'Unknown error'
    result.verdict = 'ERROR'
  }

  return result
}

function normalizeUrl(raw: string) {
  const trimmed = raw.trim()
  if (!trimmed) {
    return { valid: false, url: trimmed, error: 'Empty URL' }
  }

  const candidate = /^https?:\/\//i.test(trimmed) ? trimmed : `https://${trimmed}`

  try {
    const parsed = new URL(candidate)
    if (!['http:', 'https:'].includes(parsed.protocol)) {
      return { valid: false, url: candidate, error: `Unsupported scheme '${parsed.protocol.replace(':', '')}'` }
    }

    if (!parsed.hostname) {
      return { valid: false, url: candidate, error: 'Missing host' }
    }

    return { valid: true, url: parsed.toString(), error: null }
  } catch {
    return { valid: false, url: candidate, error: 'Malformed URL' }
  }
}

async function httpProbe(inputUrl: string): Promise<ProbeResult> {
  const redirectChain: string[] = []
  const start = Date.now()
  let currentUrl = inputUrl

  for (let i = 0; i <= MAX_REDIRECTS; i += 1) {
    const response = await requestOnce(currentUrl)
    const statusCode = response.statusCode ?? null

    if (statusCode && statusCode >= 300 && statusCode < 400 && response.headers.location) {
      const nextUrl = new URL(response.headers.location, currentUrl).toString()
      redirectChain.push(currentUrl)
      currentUrl = nextUrl
      continue
    }

    return {
      statusCode,
      finalUrl: currentUrl,
      redirectChain,
      responseTimeMs: Number((Date.now() - start).toFixed(1)),
      headers: response.headers,
    }
  }

  throw new Error('Too many redirects')
}

function requestOnce(targetUrl: string): Promise<http.IncomingMessage> {
  return new Promise((resolve, reject) => {
    const url = new URL(targetUrl)
    const client = url.protocol === 'https:' ? https : http

    const request = client.request(
      url,
      {
        method: 'GET',
        headers: {
          'User-Agent': 'URLRecon/1.0 (security-scanner)',
        },
      },
      (response) => {
        response.resume()
        resolve(response)
      },
    )

    request.setTimeout(REQUEST_TIMEOUT_MS, () => {
      request.destroy(new Error('Request timed out'))
    })

    request.on('error', reject)
    request.end()
  })
}

async function sslCheck(targetUrl: string) {
  const url = new URL(targetUrl)
  const host = url.hostname
  const port = url.port ? Number(url.port) : 443

  return new Promise<{
    sslValid: boolean | null
    sslDaysLeft: number | null
    sslIssuer: string | null
    error: string | null
  }>((resolve) => {
    const socket = tls.connect(
      {
        host,
        port,
        servername: host,
        rejectUnauthorized: true,
        timeout: REQUEST_TIMEOUT_MS,
      },
      () => {
        const cert = socket.getPeerCertificate()
        socket.end()

        if (!cert || !cert.valid_to) {
          resolve({
            sslValid: null,
            sslDaysLeft: null,
            sslIssuer: null,
            error: null,
          })
          return
        }

        const expiry = new Date(cert.valid_to)
        const daysLeft = Math.floor((expiry.getTime() - Date.now()) / (1000 * 60 * 60 * 24))
        const issuerValue = typeof cert.issuer === 'object'
          ? cert.issuer.O ?? cert.issuer.CN ?? 'Unknown'
          : 'Unknown'
        const issuer = Array.isArray(issuerValue) ? issuerValue.join(', ') : issuerValue

        resolve({
          sslValid: daysLeft > 0,
          sslDaysLeft: daysLeft,
          sslIssuer: issuer,
          error: null,
        })
      },
    )

    socket.on('error', (error) => {
      resolve({
        sslValid: false,
        sslDaysLeft: null,
        sslIssuer: null,
        error: `SSL error: ${error.message}`,
      })
    })

    socket.on('timeout', () => {
      socket.destroy()
      resolve({
        sslValid: false,
        sslDaysLeft: null,
        sslIssuer: null,
        error: 'SSL error: Request timed out',
      })
    })
  })
}

function classifyVerdict(result: ScanResult): Verdict {
  if (result.status_code === null) {
    return 'ERROR'
  }

  let score = 0

  if (result.status_code >= 500) {
    score += 2
  } else if (result.status_code >= 400) {
    score += 1
  }

  if (result.ssl_valid === false) {
    score += 3
  } else if (result.ssl_days_left !== null && result.ssl_days_left < 30) {
    score += 2
  }

  score += result.missing_headers.length

  if (result.redirect_chain.length > 0 && result.final_url) {
    const originalHost = new URL(result.url).host
    const finalHost = new URL(result.final_url).host
    if (originalHost !== finalHost) {
      score += 2
    }
  }

  if (score === 0) {
    return 'CLEAN'
  }

  if (score <= 3) {
    return 'WARNING'
  }

  return 'CRITICAL'
}

File: api/scan.py
import json
import ssl
import socket
import time
from datetime import datetime, timezone
from urllib.parse import urlparse
import re

try:
    import requests as req_lib
    HAS_REQUESTS = True
except ImportError:
    HAS_REQUESTS = False

# ── Vercel serverless handler ─────────────────────────────────────────────────

def handler(request):
    """Vercel Python serverless function entry point."""
    from http.server import BaseHTTPRequestHandler
    import urllib.parse

    if request.method == "OPTIONS":
        return Response("", 204, {
            "Access-Control-Allow-Origin": "*",
            "Access-Control-Allow-Methods": "POST, OPTIONS",
            "Access-Control-Allow-Headers": "Content-Type",
        })

    if request.method != "POST":
        return Response(json.dumps({"error": "POST only"}), 405, cors_headers())

    try:
        body = request.body
        if isinstance(body, bytes):
            body = body.decode()
        data = json.loads(body)
        urls = data.get("urls", [])
        if not urls:
            return Response(json.dumps({"error": "No URLs provided"}), 400, cors_headers())

        # Limit to 10 per request
        urls = urls[:10]
        results = [scan_url(u) for u in urls]
        return Response(json.dumps({"results": results}), 200, cors_headers())

    except Exception as e:
        return Response(json.dumps({"error": str(e)}), 500, cors_headers())


def cors_headers():
    return {
        "Content-Type": "application/json",
        "Access-Control-Allow-Origin": "*",
    }


class Response:
    def __init__(self, body, status=200, headers=None):
        self.body = body
        self.status_code = status
        self.headers = headers or {}


# ── URL validation ────────────────────────────────────────────────────────────

_URL_RE = re.compile(
    r"^https?://([\w\-]+\.)+[\w\-]{2,}(:\d{1,5})?(/.*)?$",
    re.IGNORECASE,
)

def validate_url(raw: str):
    raw = raw.strip()
    if not raw:
        return False, raw, "Empty URL"
    if not raw.startswith(("http://", "https://")):
        raw = "https://" + raw
    parsed = urlparse(raw)
    if parsed.scheme not in ("http", "https"):
        return False, raw, f"Unsupported scheme '{parsed.scheme}'"
    if not parsed.netloc:
        return False, raw, "Missing host"
    if not _URL_RE.match(raw):
        return False, raw, "Malformed URL"
    return True, raw, None


# ── Security headers audit ────────────────────────────────────────────────────

SECURITY_HEADERS = [
    "Strict-Transport-Security",
    "Content-Security-Policy",
    "X-Frame-Options",
    "X-Content-Type-Options",
    "Referrer-Policy",
    "Permissions-Policy",
]


# ── Core scanner ──────────────────────────────────────────────────────────────

def scan_url(raw_url: str) -> dict:
    result = {
        "url": raw_url,
        "valid": False,
        "status_code": None,
        "final_url": None,
        "redirect_chain": [],
        "response_time_ms": None,
        "ssl_valid": None,
        "ssl_days_left": None,
        "ssl_issuer": None,
        "missing_headers": [],
        "present_headers": [],
        "verdict": "ERROR",
        "error": None,
    }

    valid, normalised, err = validate_url(raw_url)
    if not valid:
        result["error"] = err
        return result

    result["valid"] = True
    result["url"] = normalised

    try:
        _http_probe(result)
        if result["final_url"] and result["final_url"].startswith("https://"):
            _ssl_check(result)
        _classify_verdict(result)
    except Exception as e:
        result["error"] = str(e)
        result["verdict"] = "ERROR"

    return result


def _http_probe(result: dict):
    if not HAS_REQUESTS:
        raise RuntimeError("requests library not available")

    session = req_lib.Session()
    session.headers["User-Agent"] = "URLRecon/1.0 (security-scanner)"

    start = time.monotonic()
    resp = session.get(result["url"], timeout=8, allow_redirects=True)
    elapsed = (time.monotonic() - start) * 1000

    result["status_code"] = resp.status_code
    result["final_url"] = resp.url
    result["response_time_ms"] = round(elapsed, 1)
    result["redirect_chain"] = [r.url for r in resp.history]

    headers_lower = {k.lower() for k in resp.headers}
    for h in SECURITY_HEADERS:
        if h.lower() in headers_lower:
            result["present_headers"].append(h)
        else:
            result["missing_headers"].append(h)


def _ssl_check(result: dict):
    host = urlparse(result["final_url"]).hostname
    port = urlparse(result["final_url"]).port or 443

    ctx = ssl.create_default_context()
    try:
        with socket.create_connection((host, port), timeout=8) as sock:
            with ctx.wrap_socket(sock, server_hostname=host) as ssock:
                cert = ssock.getpeercert()

        expire_str = cert.get("notAfter", "")
        if expire_str:
            expire_dt = datetime.strptime(expire_str, "%b %d %H:%M:%S %Y %Z")
            expire_dt = expire_dt.replace(tzinfo=timezone.utc)
            days_left = (expire_dt - datetime.now(timezone.utc)).days
            result["ssl_days_left"] = days_left
            result["ssl_valid"] = days_left > 0

        issuer_pairs = dict(x[0] for x in cert.get("issuer", []))
        result["ssl_issuer"] = issuer_pairs.get("organizationName", "Unknown")

    except ssl.SSLCertVerificationError as e:
        result["ssl_valid"] = False
        result["error"] = f"SSL error: {e}"


def _classify_verdict(result: dict):
    score = 0
    if result["status_code"] is None:
        result["verdict"] = "ERROR"
        return
    if result["status_code"] >= 500:
        score += 2
    elif result["status_code"] >= 400:
        score += 1

    if result["ssl_valid"] is False:
        score += 3
    elif result["ssl_days_left"] is not None and result["ssl_days_left"] < 30:
        score += 2

    score += len(result["missing_headers"])

    if result["redirect_chain"]:
        orig = urlparse(result["url"]).netloc
        final = urlparse(result["final_url"]).netloc
        if orig and final and orig != final:
            score += 2

    if score == 0:
        result["verdict"] = "CLEAN"
    elif score <= 3:
        result["verdict"] = "WARNING"
    else:
        result["verdict"] = "CRITICAL"
